## Materials and Methods {#methods}

All analyses presented in this paper can be reproduced and inspected using the Jupyter (version XX) [citation] notebook (.ipynb) within the associated github repository (https://github.com/fmaguire/MAG\_gi\_plasmid\_analysis).
The specific code version used for this paper is also archived under DOI:XXXYYY.

### Metagenome Simulation

All genomes were selected from the completed RefSeq genomes as of date XX/YY.  
Genomic islands for these genomes were previously predicted using IslandPath-DIMOB [citations] and collated into the IslandViewer database (url) [citation].
Plasmid sequences and numbers were recovered for each genome using the linked genbank Project IDs.
30 genomes were arbitrarily chosen to exemplify the following criteria:
    - 10 with high numbers of plasmids
    - 10 with a very high proportion (\>X%) of chromosomes corresponding to composition detected GIs.
    - 10 with a very low proportion (\<Y%) of chromosomes corresponding to composition detected GIs.
The data used to select the taxa is listed in `data/gi_plasmid_analysis.tsv` and the details of the select taxa are listed in `data/taxa_metadata.tsv`
The sequences themselves are in `data/sequences/sequences.tar.bz2`

[TABLE OF GENOMES AND SUMMARY]
PROJECTID TAXA PROPGI PLASMIDS RELABUNDANCEVAL PLASMIDCOPYNUMBER CRITERIA 

In accordance to the recommendation in the CAMI pre-print (DOI:XXX CITATION) the genomes were assigned a relative abundance following a log-normal distribituion (\mu = 1, \sigma = 2).
Plasmid copy number could not be accurately found in any database, therefore, plasmids were assigned a copy number regime: low (1-20), medium (20-100), or high (500-1000) at a 2:1:1 rate.
Within each regime the exact copy number was selected using an appropriately scaled gamma distribution (\alpha = 4, \beta = 1) or the minimum edge of the regime.
Finally, the effective plasmid relative abundance was determined by multiplying the plasmid copy number with the genome relative abundance.

Sequences were then concatenated into a single fasta file with the appropriate relative abudance. 
MiSeq v3 250bp paired-end reads with a mean fragment length of 1000bp (standard deviation of 50bp) were then simulated using art\_illumina (version XX) [citation] at a fold coverage of 2.9.
These resulted in 31

The selection of relative abundance and metagenome simulation itself was performed using the `data_simluation/simulate_metagenome.py` script.

## Metagenome Assembled Genome Pipeline

Use approach specified in Laura Hug's MAG textbook chapter

- Sickle quality control

- fq2fa in idba\_ud sampler to interleave paired reads (drop singletons) 

- subsample 5%, 10%, 25%, 33%, 66%, and 100% of read

- assemble each with idba\_ud

- map reads to contigs for each subsample with with `Bowtie2` and assess coverage of each contig with `BEDtools`

- merge all assemblies with Minimus2 (increasing overlap to 200 and increase minimum percent identity for overlap to 97%)

- Metabat and CONCOCT to do binning and compare
 (Kang et al. provide a
reasonable comparison of these methods)

## MAG assessment

- Annotate recovered genomes and original genomes with prokka on Vanilla

- predict GI from recovered genome with islandpath-DIMOB

- create list of genes in GI and non-GI of each genome

- look at proportion of missed genes in GI and non-GI of each genome



### MAG Recovery


### MAG Assessment

