## Materials and Methods {#methods}

All analyses presented in this paper can be reproduced and inspected using the Jupyter (version XX) [citation] notebook (.ipynb) within the associated github repository (https://github.com/fmaguire/MAG\_gi\_plasmid\_analysis).
The specific code version used for this paper is also archived under DOI:XXXYYY.

### Metagenome Simulation

All genomes were selected from the set of completed RefSeq genomes as of April 2019.  
Genomic islands for these genomes were previously predicted using IslandPath-DIMOB [@doi:10.1093/bioinformatics/bty095] and collated into the IslandViewer database (http://www.pathogenomics.sfu.ca/islandviewer) [@doi:10.1093/nar/gkv401].
Plasmid sequences and numbers were recovered for each genome using the linked genbank Project IDs.

30 genomes were arbitrarily chosen to exemplify the following criteria:
    - 10 with high numbers of plasmids
    - 10 with a very high proportion (\>X%) of chromosomes corresponding to composition detected GIs.
    - 10 with a very low proportion (\<Y%) of chromosomes corresponding to composition detected GIs.

The data used to select the taxa is listed in Supplemental Table 1 and the details of the selected subset taxa are listed in Supplemental Table 2 with their NCBI accessions.
The sequences themselves are in `data/sequences/sequences.tar.bz2`

In accordance to the recommendation in the CAMI challenge [@doi:10.1038/nmeth.4458] the genomes were randomly assigned a relative abundance following a log-normal distribituion (\mu = 1, \sigma = 2).
Plasmid copy number estimates could not be accurately found for all organisms, therefore, plasmids were randomly assigned a copy number regime: low (1-20), medium (20-100), or high (500-1000) at a 2:1:1 rate.
Within each regime the exact copy number was selected using an appropriately scaled gamma distribution (\alpha = 4, \beta = 1) or the minimum edge of the regime.
Finally, the effective plasmid relative abundance was determined by multiplying the plasmid copy number with the genome relative abundance.
The full set of randomly assigned relative abundances and copy numbers can be found in Supplemental Table 3.

Sequences were then concatenated into a single fasta file with the appropriate relative abudance. 
MiSeq v3 250bp paired-end reads with a mean fragment length of 1000bp (standard deviation of 50bp) were then simulated using art\_illumina (v2016.06.05) [@doi:10.1093/bioinformatics/btr708] at a fold coverage of 2.9 resulting in a simulate metagenome of 31,174,411 read pairs.

The selection of relative abundance and metagenome simulation itself was performed using the `data_simluation/simulate_metagenome.py` script.

### Metagenome Assembled Genome Recovery

Reads were trimmed using sickle (v1.33) [@doi:] resulting in 25,682,644 surviving read pairs.
The trimmed reads were then assembled using 3 different metagenomic assemblers: metaSPAdes (v3.13.0)[@doi:10.1101/gr.213959.116] , IDBA-UD (v1.1.3) [@doi:10.1093/bioinformatics/bts174], and megahit (v1.1.3) [@doi:10.1093/bioinformatics/btv033]).
The resulting assemblies were summarised using metaQUAST (v5.0.2) [@doi:10.1093/bioinformatics/btv697].
The assemblies were indexed and reads mapped back using Bowtie 2 (v2.3.4.3) [@doi:10.1038/nmeth.1923].

Samtools (v1.9) were then used to sort the read mappings and the read coverage calculated using the MetaBAT 2 accessory script (jgi\_summarize\_bam\_contig\_depths. 
The 3 metagenome assemblies were then separately binned using CONCOCT (v0.4.2) [@doi:10.1093/bioinformatics/btw290], MetaBAT 2 (v2.13) [@doi:10.7287/peerj.preprints.27522v1], and MaxBin 2 (v2.2.6) [@doi:10.1093/bioinformatics/btv638]. 
As per the specified manual instructions CONCOCT used a slightly different approach to estimate read coverage.
The supplied accessory scripts were also used to cut contigs into 10 kilobase fragments (cut\_up\_fasta.py) and read coverage calculated for the fragments (concoct\_coverage\_table.py).
These fragment coverages were then used to bin the 10kb fragments before the clustered fragments were merged (merge\_cutup\_clustering.py) to create the final CONCOCT MAG bins (extra\_fasta\_bins.py)
Finally, for each metagenome assembly the predicted bins from these 3 binners were combined using DAS Tool (v1.1.1) [@doi:10.1038/s41564-018-0171-1].
This resulted in 12 separate sets of MAGs (one set for each assembler and binner pair).

### MAG assessment


// (Kang et al. provide a
//reasonable comparison of these methods)

